tests/aws/services/events/test_events_targets.py

```diff
@@ -33,7 +34,199 @@
+            actions="execute-api:Invoke",
+            assume_policy_doc=json.dumps(assume_role_policy_document),
+            resource=source_arn,
+            attach=False  # Since we're using put_role_policy, not attach_role_policy
+        )
+
+        # Allow some time for IAM role propagation (only needed in AWS)
+        if is_aws_cloud():
+            time.sleep(10)
+
+        # Step e: Add the API Gateway as a target with the RoleArn
+        target_id = f"target-{short_uid()}"
+
+        # Construct the API target ARN
+        api_target_arn = f'arn:aws:execute-api:{region_name}:{account_id}:{api_id}/{stage_name}/POST/test'
+
+        put_targets_response = aws_client.events.put_targets(
+            Rule=rule_name,
+            EventBusName=event_bus_name,
+            Targets=[
+                {
+                    "Id": target_id,
+                    "Arn": api_target_arn,
+                    "RoleArn": role_arn,
+                    "Input": json.dumps({"message": "Hello from EventBridge"}),
+                    "RetryPolicy": {
+                        'MaximumRetryAttempts': 0
+                    }
+                }
+            ]
+        )
+        snapshot.match("put_targets_response", put_targets_response)
+        assert put_targets_response['FailedEntryCount'] == 0
+
+        # Step f: Send an event to EventBridge
+        event_entry = {
+            "EventBusName": event_bus_name,
+            "Source": "test.source",
+            "DetailType": "test.detail.type",
+            "Detail": json.dumps({"message": "Hello from EventBridge"})
+        }
+        put_events_response = aws_client.events.put_events(
+            Entries=[event_entry]
+        )
+        snapshot.match("put_events_response", put_events_response)
+        assert put_events_response['FailedEntryCount'] == 0
+
+        # Step g: Verify the Lambda invocation
+        events = retry(
+            check_expected_lambda_log_events_length,
+            retries=10,
+            sleep=10,
+            sleep_before=10,
+            function_name=function_name,
+            expected_length=1,
+            logs_client=aws_client.logs,
+        )
+        snapshot.match("lambda_logs", events)
+
+        # Step h: Clean up resources
+        cleanups.append(lambda: aws_client.events.remove_targets(Rule=rule_name, EventBusName=event_bus_name, Ids=[target_id]))
+        cleanups.append(lambda: aws_client.events.delete_rule(Name=rule_name, EventBusName=event_bus_name))
+        cleanups.append(lambda: aws_client.events.delete_event_bus(Name=event_bus_name))
+        cleanups.append(lambda: aws_client.apigateway.delete_rest_api(restApiId=api_id))
+        cleanups.append(lambda: aws_client.lambda_.delete_function(FunctionName=function_name))
+        cleanups.append(lambda: aws_client.iam.delete_role_policy(RoleName=role_name, PolicyName=f'p-{role_name}'))
+        cleanups.append(lambda: aws_client.iam.delete_role(RoleName=role_name))
```

Comment:
To clarify my previous comment:
- These appends (where not handled by a fixture) should take place _directly after creating the resource in question_ since the `cleanups` fixture is executed regardless of whether the test fails or passes.
- If you are using a fixture to create a resource (i.e `create_lambda_function`) I believe it is safe to assume there are already some predefined delete operations that have been defined within the fixture that pytest will execute upon teardown.

Let's compare the following cases:

#### Create, Assert, Append Cleanup

If the assertion fails, the resource will not ever be cleaned up, since we teardown/exit the test before passing that delete operation to our cleanups fixture:

```python3
aws_client.lambda_create_function(functionName="my-awesome-function")
assert put_events_response['FailedEntryCount'] == 0
cleanups.append(lambda: aws_client.lambda_.delete_function("my-awesome-function")
```

#### Create, Append Cleanup, Assert

In the below case, if the assertion fails, the resource will be cleaned up since we have passed the delete operation to the cleanups fixture:

```python3
aws_client.lambda_create_function(functionName="my-awesome-function")
cleanups.append(lambda: aws_client.lambda_.delete_function("my-awesome-function")
assert put_events_response['FailedEntryCount'] == 0
```

Reply:
as I am using fixtures for all the resources so is it safe to say that I don't need the `cleanup`? I have actually tested it by removing the cleanups.

Reply:
Yeah I think it's safe to remove the cleanups